File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/morrisTraversalPreorder.ts
import {arrayToBTLevelOrder, BTNode} from "./BinaryTreeOperations";

function getPreorder(root: BTNode) {
	const preorder = [];
	let cur = root;
	while (cur) {
		if (!cur.left) {
			preorder.push(cur.data);
			cur = cur.right;
		} else {
			let prev = cur.left;
			while (prev.right && prev.right != cur) {
				prev = prev.right;
			}
			if (!prev.right) {
				prev.right = cur;
				preorder.push(cur.data);
				cur = cur.left;
			} else {
				prev.right = null;
				cur = cur.right;
			}
		}
	}
	console.log(preorder)
}

let root = arrayToBTLevelOrder([1,2,3,4,5,6,7,8])
getPreorder(root);

File: StriverA2Z/Java/BinaryTrees/MorrisTraversalPreorder.java
import java.util.ArrayList;

public class MorrisTraversalPreorder {
    public static void main(String[] args) {
        Node root = BinaryTreeOperations.arrayToBTLevelOrder(new Integer[]{1, 2, 3, 4, 5, 6, 7, 8});
        getPreorder(root);
    }

    public static void getPreorder(Node root) {
        ArrayList<Integer> preorder = new ArrayList<>();
        Node cur = root;
        while (cur != null) {
            if (cur.left == null) {
                preorder.add(cur.data);
                cur = cur.right;
            } else {
                Node prev = cur.left;
                while (prev.right != null && prev.right != cur) {
                    prev = prev.right;
                }
                if (prev.right == null) {
                    prev.right = cur;
                    preorder.add(cur.data);
                    cur = cur.left;
                } else {
                    prev.right = null;
                    cur = cur.right;
                }
            }
        }
        for (int i : preorder) {
            System.out.print(i + " ");
        }
    }
}
